io_gripper_controller/src/io_gripper_controller.cpp

```diff
@@ -0,0 +1,1379 @@
+  {
+    case gripper_state_type::IDLE:
+      // do nothing
+      break;
+    case gripper_state_type::SET_BEFORE_COMMAND:
+      for (size_t i = 0; i < set_before_command_open.size(); ++i)
+      {
+        setResult =
+          find_and_set_command(set_before_command_open[i], set_before_command_open_values[i]);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to set the command state for %s",
+            set_before_command_open[i].c_str());
+        }
+      }
+
+      gripper_state_buffer_.writeFromNonRT(gripper_state_type::OPEN_GRIPPER);
+      break;
+    case gripper_state_type::OPEN_GRIPPER:
+      // now open the gripper
+      for (size_t i = 0; i < command_ios_open.size(); ++i)
+      {
+        setResult = find_and_set_command(command_ios_open[i], command_ios_open_values[i]);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to set the command state for %s",
+            command_ios_open[i].c_str());
+        }
+      }
+
+      gripper_state_buffer_.writeFromNonRT(gripper_state_type::CHECK_GRIPPER_STATE);
+      break;
+    case gripper_state_type::CHECK_GRIPPER_STATE:
+      // check the state of the gripper
+      check_state_ios_ = false;
+      for (size_t i = 0; i < state_ios_open.size(); ++i)
+      {
+        setResult = find_and_get_state(state_ios_open[i], state_value_);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to get the state for %s", state_ios_open[i].c_str());
+        }
+        else
+        {
+          if (abs(state_value_ - state_ios_open_values[i]) < std::numeric_limits<double>::epsilon())
+          {
+            check_state_ios_ = true;
+          }
+          else
+          {
+            check_state_ios_ = false;
+            break;
+          }
+        }
+      }
+      if (check_state_ios_)
+      {
+        gripper_state_buffer_.writeFromNonRT(gripper_state_type::SET_AFTER_COMMAND);
+      }
+      break;
+    case gripper_state_type::SET_AFTER_COMMAND:
+      for (size_t i = 0; i < set_after_command_open.size(); ++i)
+      {
+        setResult =
+          find_and_set_command(set_after_command_open[i], set_after_command_open_values[i]);
+        if (!setResult)
+        {
+          RCLCPP_ERROR(
+            get_node()->get_logger(), "Failed to set the command state for %s",
+            set_after_command_open[i].c_str());
+        }
+      }
+      for (size_t i = 0; i < params_.open.joint_states.size(); ++i)
+      {
+        joint_state_values_[i] = params_.open.joint_states[i];
+      }
+      gripper_state_buffer_.writeFromNonRT(gripper_state_type::IDLE);
+      openFlag_.store(false);
+      gripper_service_buffer_.writeFromNonRT(service_mode_type::IDLE);
+      break;
+    default:
+      break;
+  }
+}
+
+void IOGripperController::handle_reconfigure_state_transition(const reconfigure_state_type & state)
```

Comment:
Proposal here. Can we do this general check only when activating the controller, i.e., `on_activate` because later when we're switching states we know already which states do we expect, and then we check only those.

The main idea is to avoid so many iterations in every control step is this is not even necessary.

Reply:
Can we also please add tests with different initial states of the gripper? To make sure that we recognize it correctly.